106452 implement va notify callback

Khoa-V-Nguyen opened the pull request:
Summary
- Feature toggle (Flipper): ✅ Yes – accredited_representative_portal_email_delivery_callback
- Added a custom callback handler (VANotify::EmailDeliveryStatusCallback) to track and log delivery statuses for outgoing VA Notify emails.
- This is a feature addition, not a bug fix — focused on enhancing observability and error tracking.
- The callback leverages callback_klass and callback_metadata to:
- Log delivery event metadata
- Increment StatsD metrics across all delivery statuses (e.g., delivered, failure, unknown)
- The implementation is gated behind a Flipper flag to allow gradual rollout and testing.
- Team: FAR — owns the implementation and supporting jobs/controllers.

## Related issue(s)

- https://github.com/department-of-veterans-affairs/va.gov-team/issues/106452

## Testing done

- [x] *New code is covered by unit tests*
- Behavior prior to change: no VA Notify callback logic existed for tracking email delivery events.
- Testing steps:
  - Run `EmailDeliveryStatusCallback` specs and verify logs + StatsD metrics match delivery states.
  - Run `PowerOfAttorneyRequestEmailJob` and `NextStepsEmailController` specs to confirm callback metadata is passed to the service layer.
  - Confirm delivery events for `delivered`, `permanent-failure`, and unexpected statuses are logged appropriately.
- *Not behind a flipper* — tested as-is in staging using real notification IDs and by invoking the callback manually in `rails c`.

## What areas of the site does it impact?

- Affects backend delivery logic for email notifications triggered from:
  - `NextStepsEmailController`
  - `PowerOfAttorneyRequestEmailJob`
- No frontend changes, but logs/metrics now reflect delivery outcomes of VA Notify emails.

## Acceptance criteria

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No error nor warning in the console.
- [x] Events are being sent to the appropriate logging solution
- [x] Documentation has been updated (link to documentation) — _internal Slack threads and PR-level docs_
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Feature/bug has a monitor built into Datadog (if applicable) — _logs and StatsD metrics are emitted_
- [x] If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected

Khoa-V-Nguyen committed:
feat: add EmailDeliveryStatusCallback for VANotify delivery tracking

- Implements VANotify::EmailDeliveryStatusCallback class
- Handles delivered, permanent-failure, temporary-failure, and unknown statuses
- Emits StatsD metrics and logs structured metadata
- Includes corresponding RSpec coverage

This callback is meant to be passed via callback_klass to VANotify::Service and tracks email status for downstream visibility.

Khoa-V-Nguyen committed:
feat: wire VANotify callback into NextStepsEmailController

- Adds `email_delivery_callback` method returning proper callback_klass and metadata
- Passes callback options to VANotify::EmailJob
- Ensures emails from Representation Management are tracked using EmailDeliveryStatusCallback
- Updates test to verify VANotify::EmailJob receives correct callback args

Khoa-V-Nguyen committed:
feat: add callback support to PowerOfAttorneyRequestEmailJob

- Injects EmailDeliveryStatusCallback via callback options into VaNotify::Service
- Ensures callback_klass and statsd_tags are properly passed
- Adds spec verifying VaNotify::Service receives the expected callback options

This change aligns Accredited Representative Portal emails with centralized callback monitoring via EmailDeliveryStatusCallback.

opticbob reviewed (dismissed):
Looks good!  I like how the statsd logging has fallbacks if it's missing.

holdenhinkle commented on modules/va_notify/app/services/va_notify/email_delivery_status_callback.rb:
```diff
@@ -0,0 +1,61 @@
+# frozen_string_literal: true
+
+module VANotify
```
Nice of you to create `VANotify::EmailDeliveryStatusCallback` for everyone!

If the VANotify team is cool with this they should update their docs by pointing to this.

holdenhinkle commented on modules/accredited_representative_portal/app/sidekiq/accredited_representative_portal/power_of_attorney_request_email_job.rb:
```diff
@@ -69,5 +69,17 @@ def generate_personalisation(notification)
+        callback_metadata: {
+          statsd_tags: {
+            service: 'accredited-representative-portal',
+            function: 'poa_request_submission'
```
Should the `function` value be dynamic?

This job seems to handle emails for the scenarios listed in `#generate_personalisation`:

```
    def generate_personalisation(notification)
      personalisation_klass =
        case notification.type
        when 'requested'
          EmailPersonalisations::Requested
        when 'declined'
          EmailPersonalisations::Declined
        when 'expiring'
          EmailPersonalisations::Expiring
        when 'expired'
          EmailPersonalisations::Expired
        end

      personalisation_klass.generate(
        notification
      )
    end
 ```

Khoa-V-Nguyen replied:
Modified `email_delivery_callback` to embed notification type in `function` tag

holdenhinkle reviewed (commented):
I left a comment which pertains to [the ticket's description](https://github.com/department-of-veterans-affairs/va.gov-team/issues/106452).

`AccreditedRepresentativePortal::PowerOfAttorneyRequestEmailJob` seems to handle the following emails:

```
Appoint a Representative v2 - Digital Submit Confirm
Appoint a Representative v2 - Digital Submit Decline
Appoint a Representative v2 - Expiration Warning
Appoint a Representative v2 - Expiration Confirmation
```

So I _think_ the value for `callback_metadata.statsd_tags.function` in `#email_delivery_callback` needs to be dynamic based on the email that's being sent (e.g. submit, decline, expiration warning, expiration confirmation).

Khoa-V-Nguyen committed:
Update VANotify callback to include statsd_tags for silent failure tracking

- Modified `email_delivery_callback` to embed notification type in `function` tag
- Ensures StatsD logs include accurate context for Datadog monitoring

holdenhinkle reviewed (commented):
I think we're good with logging silent failures.

Looking at [Logging resolved failures](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/platform/practices/zero-silent-failures/logging-silent-failures.md#logging-resolved-failures) it says:

> In future, teams will be notified when VA Notify has a call back mechanism to determine if an email reached the intended recipient. At that time, teams should use the following log message:
> 
> ```
> tags = ['service:veteran-facing-forms', 'function: form submission to Lighthouse']
> StatsD.increment('silent_failure_avoided', tags:)
> ```

Just one more comment. This one is about ensuring Datadog tag values are strings and not symbols.

Khoa-V-Nguyen committed:
Ensure statsd_tags keys and values are strings in VANotify callback

- Updated `EmailDeliveryStatusCallback.extract_tags` to stringify keys and values in `statsd_tags`
- Added fallback to default tags when `statsd_tags` is missing or empty
- Ensures compatibility with Datadog which expects tag values as strings

Khoa-V-Nguyen commented on modules/va_notify/app/services/va_notify/email_delivery_status_callback.rb:
```diff
@@ -0,0 +1,61 @@
+        {}
+      end
+
+      metadata[:statsd_tags].presence || {
```
Great catch — updated the code to ensure both keys and values in statsd_tags are stringified, using to_h { |k, v| [k.to_s, v.to_s] }. This should keep everything Datadog-compliant. Let me know if you’d prefer this handled differently!

holdenhinkle reviewed (dismissed):
Great work! Thank you Khoa!

Khoa-V-Nguyen committed:
Move EmailDeliveryStatusCallback under Accredited Representative Portal module

Khoa-V-Nguyen committed:
Update files because of EmailDeliverStatusCallback was moved

holdenhinkle reviewed (dismissed):
Nice

nihil2501 commented:
@Khoa-V-Nguyen I retraced your steps in Slack and just learned that the reason `DefaultCallback` didn't do what we want is because, currently, it only carries out instrumentation for the Zero Silent Failure (ZSF) use case—which is what [these](https://github.com/department-of-veterans-affairs/vets-api/blob/8e0fbbbdcc8794a52d66e48640586f135cb7b604/modules/va_notify/lib/va_notify/default_callback.rb#L26-L32) `notification_type == 'error'` checks speak to.

Could we update this PR to include a comment explaining that this is the reason we've implemented a custom callback? Hypothetically it could mean that one day in the future we return to using `DefaultCallback` if they allow for passing other metadata that toggles on instrumentation in more scenarios. A code comment might be a useful clue towards that end.

raudabaugh commented on modules/accredited_representative_portal/app/services/email_delivery_status_callback.rb:
```diff
@@ -0,0 +1,65 @@
+
+  def self.report_success(base_metric, tags)
+    StatsD.increment("#{base_metric}.delivered", **tags)
+    StatsD.increment('silent_failure_avoided', **tags)
```
Should this only increment when `notification_type == 'error'`?

Khoa-V-Nguyen replied:
Good question — we want to track delivery status in all cases, not just when there’s an error, so we can monitor the full lifecycle and catch any silent issues proactively.

raudabaugh commented on modules/accredited_representative_portal/app/services/email_delivery_status_callback.rb:
```diff
@@ -0,0 +1,65 @@
+    StatsD.increment('silent_failure_avoided', **tags)
+  end
+
+  def self.report_failure(notification, base_metric, tags)
```
Should this also [increment the `silent_failure` metric](https://github.com/department-of-veterans-affairs/va.gov-team-sensitive/blob/master/platform/practices/zero-silent-failures/logging-silent-failures.md#logging-unresolved-silent-failures)?

Khoa-V-Nguyen commented:
> EDIT: Maybe the code comment is debatable. Either way.

Thanks for the thoughtful read-through — that makes perfect sense. I’ll add a brief comment noting why we’re using a custom callback here and the ZSF instrumentation limitation in DefaultCallback. Definitely good breadcrumbing in case we can consolidate this in the future.

Khoa-V-Nguyen committed:
refactor: move EmailDeliveryStatusCallback under AccreditedRepresentativePortal namespace

Khoa-V-Nguyen committed:
chore: add feature flag for custom email delivery callback to features.yml

Adds `accredited_representative_portal_email_delivery_callback` to control usage of custom VANotify callback for instrumentation and delivery status tracking.

Khoa-V-Nguyen committed:
feat: implement custom email delivery callback behind feature flag and update callback

- Update callback logic to `AccreditedRepresentativePortal::EmailDeliveryStatusCallback`
- Implements `accredited_representative_portal_email_delivery_callback` feature flag
- Updates `NextStepsEmailController` and `PowerOfAttorneyRequestEmailJob` to respect toggle
- Adds specs for both enabled and disabled scenarios

Khoa-V-Nguyen commented:
Had a little bit of a rebase issue. But all good now.

Khoa-V-Nguyen committed:
fix empty line

holdenhinkle reviewed (approved):
LGTM!
